scipy/spatial/setlist.pxd

```diff
@@ -47,6 +47,10 @@ cdef inline int init(setlist_t *setlist, size_t n, size_t size_guess) except -1:
         setlist.sizes[j] = 0
         setlist.alloc_sizes[j] = size_guess
         setlist.sets[j] = <int*>libc.stdlib.malloc(sizeof(int) * size_guess)
+        if setlist.sets[j] == NULL:
+            libc.stdlib.free(<void*>setlist.sets)
+            libc.stdlib.free(<void*>setlist.sizes)
+            raise MemoryError
```

Comment:
In this case, even though `init` is a `cdef` function, we *can* `raise MemoryError` here, because the `cdef` statement has the additional clause `except -1` at the end of the function definition.

There is a problem here, though.  In addition to freeing `setlist.sets` and `setlist.sizes`, we must also free `setlist.sets[k]` for `k` from `0` to` j-1`.  This must be done *before* `setlist.sets` is freed.

And while using `raise MemoryError` is consistent with the previous `raise` statements, it would be nice to add a message to all these statements, perhaps something like `raise MemoryError("failed to allocate memory in init")`.